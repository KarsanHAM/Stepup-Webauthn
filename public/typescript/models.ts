export interface SerializedPublicKeyCredentialUserEntity extends Omit<PublicKeyCredentialUserEntity, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialDescriptor extends Omit<PublicKeyCredentialDescriptor, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialCreationOptions extends Omit<PublicKeyCredentialCreationOptions, 'challenge' | 'excludeCredentials' | 'user'> {
  challenge: string;
  user: SerializedPublicKeyCredentialUserEntity;
  excludeCredentials?: SerializedPublicKeyCredentialDescriptor[];
}

export interface SerializedAuthenticatorResponse extends Omit<AuthenticatorResponse, 'clientDataJSON'> {
  clientDataJSON: string;
  attestationObject: string | null;
}

export interface SerializedPublicKeyCredential extends Omit<PublicKeyCredential, 'rawId' | 'response'> {
  rawId: string;
  response: SerializedAuthenticatorResponse;
}

export enum RegistrationState {
  DESERIALIZE_ATTESTATION_RESPONSE_OPTIONS,
  ATTESTATION_RESPONSE_OPTIONS_DE_SERIALIZED,
  REQUEST_USER_FOR_ATTESTATION,
  PUBLIC_KEY_CREDENTIALS,
  SERIALIZE_PUBLIC_KEY_CREDENTIALS,
  PUBLIC_KEY_CREDENTIALS_SERIALIZED,
  SENDING_PUBLIC_KEY_CREDENTIALS,
  RECEIVED_SERVER_RESPONSE,
  RECEIVED_SERVER_ERROR_RESPONSE,
  UNSUPORTED_PUBLIC_KEY_CREDENTIALS,
  ERROR,
}
