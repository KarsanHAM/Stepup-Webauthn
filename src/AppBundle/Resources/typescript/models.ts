export interface SerializedPublicKeyCredentialUserEntity extends Omit<PublicKeyCredentialUserEntity, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialDescriptor extends Omit<PublicKeyCredentialDescriptor, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialCreationOptions extends Omit<PublicKeyCredentialCreationOptions, 'challenge' | 'excludeCredentials' | 'user'> {
  challenge: string;
  user: SerializedPublicKeyCredentialUserEntity;
  excludeCredentials?: SerializedPublicKeyCredentialDescriptor[];
}
