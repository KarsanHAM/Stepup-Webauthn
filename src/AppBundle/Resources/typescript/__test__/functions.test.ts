import { set } from 'ramda';
import {
  Base64UrlSafeToUInt8,
  Base64UrlSafeToUInt8Id,
  Base64UrlSafeToUInt8Ids,
  deSerializedPublicKeyCredentialCreationOptions,
  idLens,
  optionalBase64UrlSafeToUInt8Ids,
} from '../functions';
import { SerializedPublicKeyCredentialCreationOptions } from '../models';

it('base64ToUInt8', () => {
  expect(Base64UrlSafeToUInt8('z4Ag4oiIIOKEnQ==')).toEqual(new Uint8Array([207, 128, 32, 226, 136, 136, 32, 226, 132, 157]));
});

it('idLens', () => {
  expect(set(idLens, 2, { id: 1, foo: 'bar' })).toEqual({
    id: 2,
    foo: 'bar',
  });
});

it('base64ToUInt8Id', () => {
  expect(Base64UrlSafeToUInt8Id({ id: 'z4Ag4oiIIOKEnQ==', foo: 'bar' })).toEqual({
    id: new Uint8Array([207, 128, 32, 226, 136, 136, 32, 226, 132, 157]),
    foo: 'bar',
  });
});

it('base64ToUInt8Ids', () => {
  expect(Base64UrlSafeToUInt8Ids([{ id: 'z4Ag4oiIIOKEnQ==', foo: 'bar' }, { id: 'NDM1MzQ1MzQ=', bar: 'foo' }])).toEqual([
    {
      id: new Uint8Array([207, 128, 32, 226, 136, 136, 32, 226, 132, 157]),
      foo: 'bar',
    },
    {
      id: new Uint8Array([52, 51, 53, 51, 52, 53, 51, 52]),
      bar: 'foo',
    },
  ]);
});

it('optionalBase64ToUInt8Ids can convert', () => {
  expect(optionalBase64UrlSafeToUInt8Ids([{ id: 'z4Ag4oiIIOKEnQ==', foo: 'bar' }, {
    id: 'NDM1MzQ1MzQ=',
    bar: 'foo',
  }])).toEqual([
    {
      id: new Uint8Array([207, 128, 32, 226, 136, 136, 32, 226, 132, 157]),
      foo: 'bar',
    },
    {
      id: new Uint8Array([52, 51, 53, 51, 52, 53, 51, 52]),
      bar: 'foo',
    },
  ]);
});

it('optionalBase64ToUInt8Ids can be empty', () => {
  expect(optionalBase64UrlSafeToUInt8Ids(undefined)).toEqual(undefined);
});

it('deSerializedPublicKeyCredentialCreationOptions', () => {
  const publicKeyOptions: SerializedPublicKeyCredentialCreationOptions = {
    rp: { name: 'My Super Secured Application', id: 'foo.example.com' },
    pubKeyCredParams: [{ type: 'public-key', alg: -7 }],
    challenge: '237glVYXXQoFNJCF_7fgCFwKneTN5QsWWR9hXT41rTY',
    attestation: 'none',
    user: {
      name: '@cypher-Angel-3000',
      id: 'MTIzZTQ1NjctZTg5Yi0xMmQzLWE0NTYtNDI2NjU1NDQwMDAw',
      displayName: 'Mighty Mike',
    },
    authenticatorSelection: { requireResidentKey: false, userVerification: 'preferred' },
    excludeCredentials: [{ type: 'public-key', id: 'QUJDREVGR0g=' }],
    extensions: { loc: true },
    timeout: 20000,
  };
  expect(deSerializedPublicKeyCredentialCreationOptions(publicKeyOptions)).toMatchSnapshot();
});
