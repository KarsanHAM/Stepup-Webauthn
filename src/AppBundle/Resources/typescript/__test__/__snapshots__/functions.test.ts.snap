// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deSerializedPublicKeyCredentialCreationOptions 1`] = `
Object {
  "attestation": "none",
  "authenticatorSelection": Object {
    "requireResidentKey": false,
    "userVerification": "preferred",
  },
  "challenge": Uint8Array [
    219,
    126,
    224,
    149,
    86,
    23,
    93,
    10,
    5,
    52,
    144,
    133,
    255,
    183,
    224,
    8,
    92,
    10,
    157,
    228,
    205,
    229,
    11,
    22,
    89,
    31,
    97,
    93,
    62,
    53,
    173,
    54,
  ],
  "excludeCredentials": Array [
    Object {
      "id": Uint8Array [
        65,
        66,
        67,
        68,
        69,
        70,
        71,
        72,
      ],
      "type": "public-key",
    },
  ],
  "extensions": Object {
    "loc": true,
  },
  "pubKeyCredParams": Array [
    Object {
      "alg": -7,
      "type": "public-key",
    },
  ],
  "rp": Object {
    "id": "foo.example.com",
    "name": "My Super Secured Application",
  },
  "timeout": 20000,
  "user": Object {
    "displayName": "Mighty Mike",
    "id": Uint8Array [
      49,
      50,
      51,
      101,
      52,
      53,
      54,
      55,
      45,
      101,
      56,
      57,
      98,
      45,
      49,
      50,
      100,
      51,
      45,
      97,
      52,
      53,
      54,
      45,
      52,
      50,
      54,
      54,
      53,
      53,
      52,
      52,
      48,
      48,
      48,
      48,
    ],
    "name": "@cypher-Angel-3000",
  },
}
`;
